infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java

```diff
@@ -168,60 +175,146 @@ public synchronized void handleMessageLayoutBootstrap(
+     *
+     * The only deviation from the original algorithm is (1) where instead of doing
+     * rank_msg ≥ rank_pro, we do rank_msg > rank_pro. This does not have any impact on
+     * correctness.
+     *
+     * Warning: Do not try to change the way the synchronization is done. Paxos is meat to be
+     * a message passing algorithm, meaning that only one message can be processed at
+     * any point in time.
+     *
+     * Please take a look at Technical Report UCAM-CL-TR-935 ISSN 1476-2986 for more information.
      *
      * @param msg corfu message containing LAYOUT_PREPARE
      * @param ctx netty ChannelHandlerContext
      * @param r   server router
      */
-    // TODO this can work under a separate lock for this step as it does not change the global
-    // components
     @ServerHandler(type = CorfuMsgType.LAYOUT_PREPARE)
     public synchronized void handleMessageLayoutPrepare(
             @NonNull CorfuPayloadMsg<LayoutPrepareRequest> msg,
             ChannelHandlerContext ctx,
             @NonNull IServerRouter r) {
 
-        // Check if the prepare is for the correct epoch
+        // Check if the prepare if we are bootstrapped.
         if (!isBootstrapped(msg, ctx, r)) {
             return;
         }
-        Rank prepareRank = new Rank(msg.getPayload().getRank(), msg.getClientID());
-        Rank phase1Rank = getPhase1Rank();
-        Layout proposedLayout = getProposedLayout();
 
-        long serverEpoch = getServerEpoch();
-        if (msg.getPayload().getEpoch() != serverEpoch) {
-            r.sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_EPOCH, serverEpoch));
-            log.trace("handleMessageLayoutPrepare: Incoming message with wrong epoch, got {}, "
-                            + "expected {}, message was: {}",
-                    msg.getPayload().getEpoch(), serverEpoch, msg);
+        // Check if the prepare is for the correct epoch.
+        if (!isSameEpoch(msg.getPayload().getEpoch(), msg, ctx, r)) {
             return;
         }
 
+        final Rank messageRank = new Rank(msg.getPayload().getRank(), msg.getClientID());
+        final Rank highestProposedRank = getProposeRank();
+        final Layout acceptedLayout = getAcceptedLayout();
+        final Rank highestAcceptedRank = Optional.ofNullable(getAcceptedRank())
+                .orElseGet(() -> new Rank(-1L, msg.getClientID()));
 
         // This is a prepare. If the rank is less than or equal to the phase 1 rank, reject.
-        if (phase1Rank != null && prepareRank.lessThanEqualTo(phase1Rank)) {
-            log.debug("handleMessageLayoutPrepare: Rejected phase 1 prepare of rank={}, "
-                    + "phase1Rank={}", prepareRank, phase1Rank);
-            r.sendResponse(ctx, msg, CorfuMsgType.LAYOUT_PREPARE_REJECT.payloadMsg(new
-                    LayoutPrepareResponse(phase1Rank.getRank(), proposedLayout)));
+        if (highestProposedRank == null || messageRank.greaterThan(highestProposedRank)) {
```

Comment:
can we put `highestProposedRank ` into Optional?

Reply:
Let me deal with Optional is a separate patch. I tried doing it, but then I realized the change percolates to quite a bit of code.